daemon/cmd/kube_proxy_replacement.go

```diff
@@ -459,6 +460,17 @@ func finishKubeProxyReplacementInit() error {
 	// | After this point, BPF NodePort should not be disabled |
 	// +-------------------------------------------------------+
 
+	// When WG & encrypt-node are on, a NodePort BPF to-be forwarded request
+	// to a remote node running a selected service endpoint must be encrypted.
+	// To make the NodePort's rev-{S,D}NAT translations to happen for a reply
+	// from the remote node, we need to attach bpf_host to the Cilium's WG
+	// netdev (otherwise, the WG netdev after decrypting the reply will pass
+	// it to the stack which drops the packet).
+	if option.Config.EnableNodePort &&
+		option.Config.EnableWireguard && option.Config.EncryptNode {
+		option.Config.AppendDevice(wgTypes.IfaceName)
```

Comment:
Good points! I think we forgot to mention that it's a temporary solution, and we'd like to introduce `from-wireguard` in a follow-up (to keep this PR size sane).

> Loading this here also means we will now perform the local pod delivery from cilium_wg0, no?

I think it should happen only in the case of BPF host routing which is currently mutually exclusive with WG (https://github.com/cilium/cilium/issues/15075).

> Doesn't that cause a path asymmetry when endpoint routes are disabled?

By saying asymmetry, do you mean that the networking stack is bypassed? Or that the eth0 is traversed only once for the reply path?

> Could we avoid loading this here if running in tunneling mode?

Unfortunately, that would break the reply packet path, as replies are coming back through `cilium_wg0`, while requests are forwarded via `cilium_vxlan` (it's a follow up to fix this asymmetry and the encryption issue).

Reply:
> I think it should happen only in the case of BPF host routing which is currently mutually exclusive with WG (https://github.com/cilium/cilium/issues/15075).

Ah, right. You mean [this check](https://github.com/cilium/cilium/blob/274614fa66b43cf5898fba1164a64b16eafb93b3/bpf/bpf_host.c#L528-L540) (note the comment there is incorrect, we should still be able to perform local delivery when endpoint routes are enabled).

> By saying asymmetry, do you mean that the networking stack is bypassed? Or that the eth0 is traversed only once for the reply path?

I meant the path lxc <> cilium_wg0 was not the same for SYN and SYN+ACK. Not a concern given the above check.

Reply:
> note the comment there is incorrect

Yikes! I think NO_REDIRECT, BPF host redirect, local redirect etc could be revisited in bpf_host :spaghetti:

Reply:
Yep. I'd get rid of `NO_REDIRECT` for a start; it's not even used consistently.

Reply:
I will remove the `NO_REDIRECT`, as I have introduced it :grimacing:

Reply:
https://github.com/cilium/cilium/pull/23076

Reply:
For the redirection mess, I think this issue covers it well - https://github.com/cilium/cilium/issues/16166.